Passenger AppPreloader reports existing ruby executables as non existing when BUNDLED WITH in Gemfile.lock differs from pre-installed

Rails Apps based on this image are not starting up showing "We're sorry, but something went wrong".

`/var/log/nginx/error.log` shows:

```
App 1977 output: /usr/local/rvm/rubies/ruby-3.1.5/bin/ruby: No such file or directory -- Passenger AppPreloader: /home/app/webapp (LoadError)
[ E 2024-06-18 00:20:11.9001 77/T29 age/Cor/App/Implementation.cpp:221 ]: Could not spawn process for application /home/app/webapp: The application process exited prematurely.
  Error ID: 4b6cb0aa
  Error details saved to: /tmp/passenger-error-O6tEuV.html

[ E 2024-06-18 00:20:11.9014 77/Tr age/Cor/Con/CheckoutSession.cpp:281 ]: [Client 10-1] Cannot checkout session because a spawning error occurred. The identifier of the error is 4b6cb0aa. Please see earlier logs for details about the error.
```

but the file actually exists:

```bash
root@05b1edfe5a34:/home/app/webapp# ls -la /usr/local/rvm/rubies/ruby-3.1.5/bin/ruby
-rwxrwxr-x 1 root root 215920 May 17 10:24 /usr/local/rvm/rubies/ruby-3.1.5/bin/ruby
```

and  is really executable:

```bash
root@05b1edfe5a34:/home/app/webapp# /usr/local/rvm/rubies/ruby-3.1.5/bin/ruby --version
ruby 3.1.5p252 (2024-04-23 revision 1945f8dc0e) [x86_64-linux]
```

This happens for images based on
* phusion/passenger-ruby31:3.0.4
* phusion/passenger-ruby31:3.0.5
* phusion/passenger-ruby31:3.0.6 / latest, both when using already-installed 3.1.5 and also when installing ruby-3.1.4 via rvm
* phusion/passenger-ruby32 (latest) when using ruby 3.2.2 (i.e. NOT the one which comes with the image but installing it via rvm during docker build.) and also with ruby 3.2.4 (the one which comes with the image)

There's no issue with the latest docker/passenger-ruby33 image when using the ruby which comes with the image. I didn't try installing a different version via rvm.

I first noticed this last Friday's nightly CI run (i.e. since 3.0.6 has been released).

It also happens for images built before which worked before but where container is recreated, i.e. already existing images stop working.

Nginx config sets `passenger_ruby /usr/bin/ruby;` but this is just a wrapper forwarding to /usr/local/rvm/wrappers/default/ruby which then references `/usr/local/rvm/gems/ruby-3.1.5/environment` which contains:

```
export PATH="/usr/local/rvm/gems/ruby-3.1.5/bin:/usr/local/rvm/gems/ruby-3.1.5@global/bin:/usr/local/rvm/rubies/ruby-3.1.5/bin:$PATH"
export GEM_HOME='/usr/local/rvm/gems/ruby-3.1.5'
export GEM_PATH='/usr/local/rvm/gems/ruby-3.1.5:/usr/local/rvm/gems/ruby-3.1.5@global'
export MY_RUBY_HOME='/usr/local/rvm/rubies/ruby-3.1.5'
export IRBRC='/usr/local/rvm/rubies/ruby-3.1.5/.irbrc'
unset MAGLEV_HOME
unset RBXOPT
export RUBY_VERSION='ruby-3.1.5'
```

It still doesn't work when setting `passenger_ruby /usr/local/rvm/rubies/ruby-3.1.5/bin/ruby;`


Even though I tested quite a few different combinations of image versions and ruby versions, I still don't know where the issue comes from. Newly created containers from images built `FROM phusion/passenger-ruby31` before last Thursday were working before but  are now **also broken** once recreated.

Yes, but I didn't find anything which helped me solving it. Maybe you see something?

![CleanShot 2024-06-18 at 22 11 43@2x](https://github.com/phusion/passenger-docker/assets/4932/dcd22390-31ea-4ad1-b0c2-85a2c7b46797)

![CleanShot 2024-06-18 at 22 12 44@2x](https://github.com/phusion/passenger-docker/assets/4932/284bdf6f-5349-4655-9bc1-b5b751afd643)

![CleanShot 2024-06-18 at 22 13 06@2x](https://github.com/phusion/passenger-docker/assets/4932/95f9d7cd-ff58-43fc-9651-330b15a8a48f)

![CleanShot 2024-06-18 at 22 13 29@2x](https://github.com/phusion/passenger-docker/assets/4932/4fd5665f-c993-4e44-b0cc-639332d2b04d)


![CleanShot 2024-06-18 at 22 18 22@2x](https://github.com/phusion/passenger-docker/assets/4932/b002ad5a-c2b9-4db6-942a-e4859f367b9f)

(environment shortened in screenshot, so I don't need to censor the screenshot again. 😉 )

### activated_gems
```
error_highlight => 0.6.0
did_you_mean => 1.6.3
tmpdir => 0.2.0
fileutils => 1.7.2
etc => 1.4.3
base64 => 0.2.0
shellwords => 0.2.0
strscan => 3.1.0
forwardable => 1.3.3
io-wait => 0.3.1
```


### load_path
```
/usr/local/rvm/gems/ruby-3.1.5/gems/error_highlight-0.6.0/lib
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib
/usr/local/rvm/gems/ruby-3.1.5/gems/tmpdir-0.2.0/lib
/usr/local/rvm/gems/ruby-3.1.5/gems/fileutils-1.7.2/lib
/usr/local/rvm/gems/ruby-3.1.5/gems/etc-1.4.3/lib
/usr/local/rvm/gems/ruby-3.1.5/extensions/x86_64-linux/3.1.0/etc-1.4.3
/usr/local/rvm/gems/ruby-3.1.5/gems/base64-0.2.0/lib
/usr/local/rvm/gems/ruby-3.1.5/gems/shellwords-0.2.0/lib
/usr/local/rvm/gems/ruby-3.1.5/gems/strscan-3.1.0/lib
/usr/local/rvm/gems/ruby-3.1.5/extensions/x86_64-linux/3.1.0/strscan-3.1.0
/usr/local/rvm/gems/ruby-3.1.5/gems/forwardable-1.3.3/lib
/usr/local/rvm/gems/ruby-3.1.5/gems/io-wait-0.3.1/lib
/usr/local/rvm/gems/ruby-3.1.5/extensions/x86_64-linux/3.1.0/io-wait-0.3.1
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/x86_64-linux
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/vendor_ruby/3.1.0
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/vendor_ruby/3.1.0/x86_64-linux
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/vendor_ruby
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0/x86_64-linux
```

### loaded_libs
```
enumerator.so
thread.rb
fiber.so
rational.so
complex.so
ruby2_keywords.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0/x86_64-linux/enc/encdb.so
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0/x86_64-linux/enc/trans/transdb.so
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0/x86_64-linux/rbconfig.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/compatibility.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/defaults.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/deprecate.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/errors.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/unknown_command_spell_checker.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/exceptions.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/basic_specification.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/stub_specification.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/platform.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/specification_record.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/util/list.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/version.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/requirement.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/specification.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/util.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/dependency.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/core_ext/kernel_gem.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0/x86_64-linux/monitor.so
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0/monitor.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/rubygems/path_support.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/error_highlight-0.6.0/lib/error_highlight/version.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/error_highlight-0.6.0/lib/error_highlight/base.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/error_highlight-0.6.0/lib/error_highlight/formatter.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/error_highlight-0.6.0/lib/error_highlight/core_ext.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/error_highlight-0.6.0/lib/error_highlight.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/version.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/core_ext/name_error.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/levenshtein.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/jaro_winkler.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/spell_checker.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/spell_checkers/name_error_checkers/variable_name_checker.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/spell_checkers/name_error_checkers.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/spell_checkers/method_name_checker.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/spell_checkers/key_error_checker.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/spell_checkers/null_checker.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/tree_spell_checker.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/spell_checkers/require_path_checker.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/spell_checkers/pattern_key_name_checker.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean/formatter.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/did_you_mean-1.6.3/lib/did_you_mean.rb
/usr/lib/ruby/vendor_ruby/phusion_passenger.rb
/usr/lib/ruby/vendor_ruby/phusion_passenger/loader_shared_helpers.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/fileutils-1.7.2/lib/fileutils.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/etc-1.4.3/lib/etc.so
/usr/local/rvm/gems/ruby-3.1.5/gems/tmpdir-0.2.0/lib/tmpdir.rb
/usr/lib/ruby/vendor_ruby/phusion_passenger/constants.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/base64-0.2.0/lib/base64.rb
/usr/lib/ruby/vendor_ruby/phusion_passenger/utils.rb
/usr/lib/ruby/vendor_ruby/phusion_passenger/utils/tmpio.rb
/usr/lib/ruby/vendor_ruby/phusion_passenger/platform_info.rb
/usr/lib/ruby/vendor_ruby/phusion_passenger/platform_info/operating_system.rb
/usr/lib/ruby/vendor_ruby/phusion_passenger/platform_info/ruby.rb
/usr/lib/ruby/vendor_ruby/phusion_passenger/platform_info/binary_compatibility.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/shellwords-0.2.0/lib/shellwords.rb
/usr/lib/ruby/vendor_ruby/phusion_passenger/utils/shellwords.rb
/home/app/.passenger/native_support/6.0.22/ruby-3.1.5-x86_64-linux/passenger_native_support.so
/usr/lib/ruby/vendor_ruby/phusion_passenger/native_support.rb
/usr/lib/ruby/vendor_ruby/phusion_passenger/preloader_shared_helpers.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/strscan-3.1.0/lib/strscan.so
/usr/local/rvm/gems/ruby-3.1.5/gems/forwardable-1.3.3/lib/forwardable/impl.rb
/usr/local/rvm/gems/ruby-3.1.5/gems/forwardable-1.3.3/lib/forwardable.rb
/usr/lib/ruby/vendor_ruby/phusion_passenger/utils/json.rb
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0/x86_64-linux/socket.so
/usr/local/rvm/gems/ruby-3.1.5/gems/io-wait-0.3.1/lib/io/wait.so
/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0/socket.rb
```


### rbconfig
```
DESTDIR = 
MAJOR = 3
MINOR = 1
TEENY = 5
PATCHLEVEL = 252
INSTALL = /usr/bin/install -c
EXEEXT = 
prefix = /usr/local/rvm/rubies/ruby-3.1.5
ruby_install_name = ruby
RUBY_INSTALL_NAME = ruby
RUBY_SO_NAME = ruby
exec = exec
ruby_pc = ruby-3.1.pc
CC_WRAPPER = 
PACKAGE = ruby
BUILTIN_TRANSSRCS =  enc/trans/newline.c
MKMF_VERBOSE = 0
MANTYPE = man
vendorarchhdrdir = /usr/local/rvm/rubies/ruby-3.1.5/include/ruby-3.1.0/vendor_ruby/x86_64-linux
sitearchhdrdir = /usr/local/rvm/rubies/ruby-3.1.5/include/ruby-3.1.0/site_ruby/x86_64-linux
rubyarchhdrdir = /usr/local/rvm/rubies/ruby-3.1.5/include/ruby-3.1.0/x86_64-linux
vendorhdrdir = /usr/local/rvm/rubies/ruby-3.1.5/include/ruby-3.1.0/vendor_ruby
sitehdrdir = /usr/local/rvm/rubies/ruby-3.1.5/include/ruby-3.1.0/site_ruby
rubyhdrdir = /usr/local/rvm/rubies/ruby-3.1.5/include/ruby-3.1.0
RUBY_SEARCH_PATH = 
UNIVERSAL_INTS = 
UNIVERSAL_ARCHNAMES = 
configure_args =  '--prefix=/usr/local/rvm/rubies/ruby-3.1.5' '--disable-cache' '--disable-install-doc' '--enable-shared'
CONFIGURE = configure
vendorarchdir = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/vendor_ruby/3.1.0/x86_64-linux
vendorlibdir = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/vendor_ruby/3.1.0
vendordir = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/vendor_ruby
sitearchdir = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0/x86_64-linux
sitelibdir = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby/3.1.0
sitedir = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/site_ruby
rubyarchdir = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0/x86_64-linux
rubylibdir = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0
ruby_version = 3.1.0
sitearch = x86_64-linux
arch = x86_64-linux
sitearchincludedir = /usr/local/rvm/rubies/ruby-3.1.5/include/x86_64-linux
archincludedir = /usr/local/rvm/rubies/ruby-3.1.5/include/x86_64-linux
sitearchlibdir = /usr/local/rvm/rubies/ruby-3.1.5/lib/x86_64-linux
archlibdir = /usr/local/rvm/rubies/ruby-3.1.5/lib/x86_64-linux
libdirname = libdir
RUBY_EXEC_PREFIX = /usr/local/rvm/rubies/ruby-3.1.5
RUBY_LIB_VERSION = 
RUBY_LIB_VERSION_STYLE = 3	/* full */
RI_BASE_NAME = ri
ridir = /usr/local/rvm/rubies/ruby-3.1.5/share/ri
rubysitearchprefix = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/x86_64-linux
rubyarchprefix = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/x86_64-linux
MAKEFILES = Makefile GNUmakefile
PLATFORM_DIR = 
THREAD_MODEL = pthread
SYMBOL_PREFIX = 
EXPORT_PREFIX = 
COMMON_HEADERS = 
COMMON_MACROS = 
COMMON_LIBS = 
MAINLIBS = -lz -lpthread -lrt -lrt -lgmp -ldl -lcrypt -lm 
ENABLE_SHARED = yes
DLDSHARED = gcc -shared
DLDLIBS = -lc
SOLIBS = -lz -lpthread -lrt -lrt -lgmp -ldl -lcrypt -lm 
LIBRUBYARG_SHARED = -Wl,-rpath,/usr/local/rvm/rubies/ruby-3.1.5/lib -L/usr/local/rvm/rubies/ruby-3.1.5/lib -lruby
LIBRUBYARG_STATIC = -Wl,-rpath,/usr/local/rvm/rubies/ruby-3.1.5/lib -L/usr/local/rvm/rubies/ruby-3.1.5/lib -lruby-static -lz -lpthread -lrt -lrt -lgmp -ldl -lcrypt -lm 
LIBRUBYARG = -Wl,-rpath,/usr/local/rvm/rubies/ruby-3.1.5/lib -L/usr/local/rvm/rubies/ruby-3.1.5/lib -lruby
LIBRUBY = libruby.so.3.1.5
LIBRUBY_ALIASES = libruby.so.3.1 libruby.so
LIBRUBY_SONAME = libruby.so.3.1
LIBRUBY_SO = libruby.so.3.1.5
LIBRUBY_A = libruby-static.a
RUBYW_INSTALL_NAME = 
rubyw_install_name = 
EXTDLDFLAGS = 
EXTLDFLAGS = 
strict_warnflags = -std=gnu99
warnflags = -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wold-style-definition -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable -Wundef
debugflags = -ggdb3
optflags = -O3 -fno-fast-math
NULLCMD = :
ENABLE_DEBUG_ENV = 
DLNOBJ = dln.o
INSTALL_STATIC_LIBRARY = no
MJIT_SUPPORT = yes
EXECUTABLE_EXTS = 
ARCHFILE = 
LIBRUBY_RELATIVE = no
EXTOUT = .ext
PREP = miniruby
CROSS_COMPILING = no
TEST_RUNNABLE = yes
rubylibprefix = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby
setup = Setup
ENCSTATIC = 
EXTSTATIC = 
SOEXT = so
TRY_LINK = 
PRELOADENV = LD_PRELOAD
LIBPATHENV = LD_LIBRARY_PATH
RPATHFLAG =  -Wl,-rpath,%1$-s
LIBPATHFLAG =  -L%1$-s
LINK_SO = 
ADDITIONAL_DLDFLAGS = 
ASMEXT = S
LIBEXT = a
DLEXT = so
LDSHAREDXX = g++ -shared
LDSHARED = gcc -shared
CCDLFLAGS = -fPIC
STATIC = 
ARCH_FLAG = 
DLDFLAGS = -Wl,--compress-debug-sections=zlib
ALLOCA = 
EGREP = /usr/bin/grep -E
GREP = /usr/bin/grep
dsymutil = 
codesign = 
cleanlibs = 
POSTLINK = :
incflags = 
WERRORFLAG = -Werror
CHDIR = cd -P
RMALL = rm -fr
RMDIRS = rmdir --ignore-fail-on-non-empty -p
RMDIR = rmdir --ignore-fail-on-non-empty
CP = cp
RM = rm -f
PKG_CONFIG = pkg-config
DOXYGEN = 
DOT = 
MAKEDIRS = /usr/bin/mkdir -p
MKDIR_P = /usr/bin/mkdir -p
INSTALL_DATA = /usr/bin/install -c -m 644
INSTALL_SCRIPT = /usr/bin/install -c
INSTALL_PROGRAM = /usr/bin/install -c
SET_MAKE = 
LN_S = ln -s
DLLWRAP = 
WINDRES = 
ASFLAGS = 
ARFLAGS = rcD 
try_header = 
CC_VERSION_MESSAGE = gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
CC_VERSION = gcc --version
MJIT_CC = /usr/bin/gcc
CSRCFLAG = 
COUTFLAG = -o 
OUTFLAG = -o 
CPPOUTFILE = -o conftest.i
GNU_LD = yes
GCC = yes
CPP = gcc -E
CXXFLAGS = 
OBJEXT = o
CPPFLAGS =   
LDFLAGS = -L. -fstack-protector-strong -rdynamic -Wl,-export-dynamic -Wl,--no-as-needed
CFLAGS = -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wold-style-definition -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable -Wundef  -fPIC
STRIP = strip -S -x
RANLIB = gcc-ranlib
OBJDUMP = objdump
OBJCOPY = :
NM = gcc-nm
LD = ld
CXX = g++
AS = as
AR = gcc-ar
CC = gcc
target_os = linux
target_vendor = pc
target_cpu = x86_64
target = x86_64-pc-linux-gnu
host_os = linux
host_vendor = pc
host_cpu = x86_64
host = x86_64-pc-linux-gnu
build_os = linux-gnu
build_vendor = pc
build_cpu = x86_64
build = x86_64-pc-linux-gnu
RUBY_VERSION_NAME = ruby-3.1.0
RUBYW_BASE_NAME = rubyw
RUBY_BASE_NAME = ruby
RUBY_PROGRAM_VERSION = 3.1.5
RUBY_API_VERSION = 3.1
HAVE_GIT = yes
GIT = git
cxxflags = 
cppflags = 
cflags = -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wold-style-definition -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable -Wundef
target_alias = 
host_alias = 
build_alias = 
LIBS = -lm 
ECHO_T = 
ECHO_N = -n
ECHO_C = 
DEFS = 
mandir = /usr/local/rvm/rubies/ruby-3.1.5/share/man
localedir = /usr/local/rvm/rubies/ruby-3.1.5/share/locale
libdir = /usr/local/rvm/rubies/ruby-3.1.5/lib
psdir = /usr/local/rvm/rubies/ruby-3.1.5/share/doc/ruby
pdfdir = /usr/local/rvm/rubies/ruby-3.1.5/share/doc/ruby
dvidir = /usr/local/rvm/rubies/ruby-3.1.5/share/doc/ruby
htmldir = /usr/local/rvm/rubies/ruby-3.1.5/share/doc/ruby
infodir = /usr/local/rvm/rubies/ruby-3.1.5/share/info
docdir = /usr/local/rvm/rubies/ruby-3.1.5/share/doc/ruby
oldincludedir = /usr/include
includedir = /usr/local/rvm/rubies/ruby-3.1.5/include
runstatedir = /usr/local/rvm/rubies/ruby-3.1.5/var/run
localstatedir = /usr/local/rvm/rubies/ruby-3.1.5/var
sharedstatedir = /usr/local/rvm/rubies/ruby-3.1.5/com
sysconfdir = /usr/local/rvm/rubies/ruby-3.1.5/etc
datadir = /usr/local/rvm/rubies/ruby-3.1.5/share
datarootdir = /usr/local/rvm/rubies/ruby-3.1.5/share
libexecdir = /usr/local/rvm/rubies/ruby-3.1.5/libexec
sbindir = /usr/local/rvm/rubies/ruby-3.1.5/sbin
bindir = /usr/local/rvm/rubies/ruby-3.1.5/bin
exec_prefix = /usr/local/rvm/rubies/ruby-3.1.5
PACKAGE_URL = 
PACKAGE_BUGREPORT = 
PACKAGE_STRING = 
PACKAGE_VERSION = 
PACKAGE_TARNAME = 
PACKAGE_NAME = 
PATH_SEPARATOR = :
SHELL = /bin/bash
UNICODE_VERSION = 13.0.0
UNICODE_EMOJI_VERSION = 13.1
platform = x86_64-linux
archdir = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0/x86_64-linux
topdir = /usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/3.1.0/x86_64-linux
```

### ruby_info
```
RUBY_VERSION = 3.1.5
RUBY_PLATFORM = x86_64-linux
RUBY_ENGINE = ruby
RubyGems version = 3.5.13
RubyGems paths = ["/usr/local/rvm/gems/ruby-3.1.5", "/usr/local/rvm/rubies/ruby-3.1.5/lib/ruby/gems/3.1.0"]
```

System-wide stats
```
------------- General -------------
Kernel version    : 5.15.0-112-generic
Uptime            : 4d 12h 26m 14s
Load averages     : 0.90%, 1.26%, 1.53%
Fork rate         : unknown

------------- CPU -------------
Number of CPUs    :   12
Average CPU usage :   9%  --   4% user,   1% nice,   4% system,  91% idle
  CPU 1           :   8%  --   4% user,   0% nice,   4% system,  92% idle
  CPU 2           :   8%  --   4% user,   0% nice,   4% system,  92% idle
  CPU 3           :   9%  --   4% user,   1% nice,   4% system,  91% idle
  CPU 4           :   9%  --   4% user,   1% nice,   4% system,  91% idle
  CPU 5           :   9%  --   4% user,   1% nice,   4% system,  91% idle
  CPU 6           :   9%  --   4% user,   1% nice,   4% system,  91% idle
  CPU 7           :   8%  --   4% user,   0% nice,   4% system,  92% idle
  CPU 8           :   8%  --   4% user,   0% nice,   4% system,  92% idle
  CPU 9           :   8%  --   4% user,   0% nice,   4% system,  92% idle
  CPU 10          :   9%  --   4% user,   1% nice,   4% system,  91% idle
  CPU 11          :   9%  --   4% user,   1% nice,   4% system,  91% idle
  CPU 12          :   8%  --   4% user,   1% nice,   4% system,  92% idle
I/O pressure      :   1%
  CPU 1           :   0%
  CPU 2           :   0%
  CPU 3           :   1%
  CPU 4           :   1%
  CPU 5           :   1%
  CPU 6           :   0%
  CPU 7           :   1%
  CPU 8           :   0%
  CPU 9           :   0%
  CPU 10          :   0%
  CPU 11          :   0%
  CPU 12          :   0%
Interference from other VMs:   0%
  CPU 1                    :   0%
  CPU 2                    :   0%
  CPU 3                    :   0%
  CPU 4                    :   0%
  CPU 5                    :   0%
  CPU 6                    :   0%
  CPU 7                    :   0%
  CPU 8                    :   0%
  CPU 9                    :   0%
  CPU 10                   :   0%
  CPU 11                   :   0%
  CPU 12                   :   0%

------------- Memory -------------
RAM total         :  64221 MB
RAM used          :  45811 MB (71%)
RAM free          :  18410 MB
Swap total        :  32767 MB
Swap used         :  16159 MB (49%)
Swap free         :  16608 MB
Swap in           : unknown
Swap out          : unknown
```



## Edit: adding native support log:

```
root@f6a9bebcdad4:/tmp# cat passenger_native_support-xt0ow.log
# current user is: app
# mkdir -p /home/app/.passenger/native_support/6.0.22/ruby-3.1.5-x86_64-linux
# cd /home/app/.passenger/native_support/6.0.22/ruby-3.1.5-x86_64-linux
# /usr/local/rvm/gems/ruby-3.1.5/wrappers/ruby /usr/share/passenger/ruby_extension_source/extconf.rb
checking for alloca.h... yes
checking for ruby/version.h... yes
checking for ruby/io.h... yes
checking for ruby/thread.h... yes
checking for ruby_version... no
checking for rb_thread_io_blocking_region() in ruby/io.h... yes
checking for rb_thread_call_without_gvl() in ruby/thread.h... yes
creating Makefile
# make
compiling /usr/share/passenger/ruby_extension_source/passenger_native_support.c
/usr/share/passenger/ruby_extension_source/passenger_native_support.c: In function ‘Init_passenger_native_support’:
/usr/share/passenger/ruby_extension_source/passenger_native_support.c:894:1: warning: old-style function definition [-Wold-style-definition]
  894 | Init_passenger_native_support() {
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/passenger/ruby_extension_source/passenger_native_support.c: At top level:
cc1: note: unrecognized command-line option ‘-Wno-self-assign’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-parentheses-equality’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-constant-logical-operand’ may have been intended to silence earlier diagnostics
linking shared-object passenger_native_support.so
```

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -3,6 +3,7 @@ Release 6.0.25 (Not yet released)
  * Fixes compilation with clang 19 (latest Fedora update) by dropping a buggy stddev function from the moving average header. Closes GH-2580.
  * [Standalone] Adds a config option to specify the stop timeout for Passenger: `--stop-timeout 120` or `PASSENGER_STOP_TIMEOUT=120`.
  * [Standalone] Changes Passenger's (not apps') start timeout to 25s (from 15s), stop timeouts default to 60s.
+ * [Ruby] Fixes an issue where Bundler would try to re-exec the process name instead of the script. Closes GH-2567 and GH-2577.
  * 
 
 

--- a/src/helper-scripts/meteor-loader.rb
+++ b/src/helper-scripts/meteor-loader.rb
@@ -183,8 +183,13 @@ module PhusionPassenger
         end
         exec("meteor run -p #{port} #{production} --settings settings.json")
       end
-      $0 = options["process_title"] if options["process_title"]
-      $0 = "#{$0} (#{pid})"
+
+      if options["process_title"] && !options["process_title"].empty?
+        rename_process "#{options["process_title"]} (#{pid})"
+      else
+        rename_process "#{$0} (#{pid})"
+      end
+
       return [pid, port]
     end
 

--- a/src/ruby_supportlib/phusion_passenger/loader_shared_helpers.rb
+++ b/src/ruby_supportlib/phusion_passenger/loader_shared_helpers.rb
@@ -304,9 +304,9 @@ module PhusionPassenger
       end
 
       if options["process_title"] && !options["process_title"].empty?
-        $0 = options["process_title"] + ": " + options["app_group_name"]
+        rename_process "#{options["process_title"]}: #{options["app_group_name"]}"
       else
-        $0 = "Passenger App: " + options["app_group_name"]
+        rename_process "Passenger App: #{options["app_group_name"]}"
       end
 
       # If we were forked from a preloader process then clear or
@@ -539,6 +539,14 @@ module PhusionPassenger
     def dump_envvars
       dir = "#{ENV['PASSENGER_SPAWN_WORK_DIR']}/envdump"
       try_write_file("#{dir}/envvars", ENV.to_a.map { |k, v| "#{k} = #{v}" }.join("\n"))
+    end
+
+    def rename_process(name)
+      if Process.respond_to?(:setproctitle)
+        Process.setproctitle(name)
+      else
+        $0 = name
+      end
     end
 
   private

--- a/src/ruby_supportlib/phusion_passenger/preloader_shared_helpers.rb
+++ b/src/ruby_supportlib/phusion_passenger/preloader_shared_helpers.rb
@@ -37,7 +37,7 @@ module PhusionPassenger
     def init(main_app)
       options = LoaderSharedHelpers.init(main_app)
 
-      $0 = "#{SHORT_PROGRAM_NAME} AppPreloader: #{options['app_root']}"
+      LoaderSharedHelpers.rename_process pre_name(options)
 
       if !Kernel.respond_to?(:fork)
         message = "Smart spawning is not available on this Ruby " +
@@ -56,7 +56,7 @@ module PhusionPassenger
       options
     end
 
-    def accept_and_process_next_client(server_socket)
+    def accept_and_process_next_client(server_socket, options)
       client = server_socket.accept
       client.binmode
       begin
@@ -75,7 +75,7 @@ module PhusionPassenger
       end
 
       if doc['command'] == 'spawn'
-        handle_spawn_command(client, doc)
+        handle_spawn_command(client, doc, options)
       else
         client.write(Utils::JSON.generate(
           :result => 'error',
@@ -94,7 +94,7 @@ module PhusionPassenger
       end
     end
 
-    def handle_spawn_command(client, doc)
+    def handle_spawn_command(client, doc, options)
       work_dir = doc['work_dir']
       LoaderSharedHelpers.record_journey_step_end('PRELOADER_PREPARATION',
         'STEP_PERFORMED', work_dir)
@@ -114,7 +114,7 @@ module PhusionPassenger
 
       if pid.nil?
         begin
-          $0 = "#{$0} (forking...)"
+          LoaderSharedHelpers.rename_process "#{pre_name(options)} (forking...)"
           LoaderSharedHelpers.record_journey_step_end('PRELOADER_FORK_SUBPROCESS',
             'STEP_PERFORMED', work_dir)
           LoaderSharedHelpers.run_block_and_record_step_progress('PRELOADER_SEND_RESPONSE', work_dir) do
@@ -164,7 +164,7 @@ module PhusionPassenger
         # https://code.google.com/p/phusion-passenger/issues/detail?id=915
         ios = Kernel.select([server_socket, STDIN])[0]
         if ios.include?(server_socket)
-          result, subprocess_work_dir = accept_and_process_next_client(server_socket)
+          result, subprocess_work_dir = accept_and_process_next_client(server_socket, options)
           if result == :forked
             return subprocess_work_dir
           end
@@ -188,6 +188,12 @@ module PhusionPassenger
         File.unlink(socket_filename) rescue nil
       end
     end
+
+    private
+    def pre_name(options)
+      "#{SHORT_PROGRAM_NAME} AppPreloader: #{options['app_root']}"
+    end
+
   end
 
 end # module PhusionPassenger

